### PR TITLE
Etag-based versioning for HTTP resource

### PR DESCRIPTION
**Pull request by DavidEBest**
This is mostly working, but I want to get some more tests in here to cover edge cases. After that, I'll submit a PR upstream.

Feedback on the approach would be appreciated.

@afeld, @ctro

### CODE REVIEW

File: assets/resource.py [hunk 1]
```diff
@@ -46,6 +47,26 @@ def check(self, source, version):
...
+        elif isinstance(ssl_verify, str):
+            verify = str(tempfile.NamedTemporaryFile(delete=False, prefix='ssl-').write(verify))
+
+        # request index
+        response = requests.request('GET', index, verify=verify)
+        response.raise_for_status()
+
+        if source.get('etag'):
+            return self.check_etag(source, version, response)
+        else:
```
Comment (ctro) [inline assets/resource.py:75]: I could probably use a comment here to explain why/when sites don't have etags, and what it means to fall back to regex version searching

**Comment by ctro** [comment]
Took a quick tour though the code and it looks good so far & makes sense to me

File: assets/resource.py (hunk 1 above)
Comment (DavidEBest) [inline assets/resource.py:75]: Makes sense. Unfortunately, both of these methods are a bit contrived, and I don't see a use case for fallback. The resource should fail if it lacks an etag. Sounds like I need another test or two. :)

**Comment by DavidEBest** [comment]
@afeld, @ctro This should be good to review now. My intended course of action is:
- Get this reviewed & merged here
- Push to docker hub in 18fgsa, so we can start using it
- Submit a PR upstream
- If/when it gets merged, switch to using it.

Does that make sense/work for you two?

I've been testing it as `davidebest/htttp-resource` for now, and it all works locally for the health-check pipeline.

File: assets/resource.py (hunk 1 above)
Comment (afeld) [inline assets/resource.py:75]: > what it means to fall back to regex version searching

Yeah, maybe something like "this checks the body of the response for something that looks like a version number"? That being said, I can't think of a real-world scenario where that would be useful, so imagine we will end up taking it out.

> I don't see a use case for fallback

Well, sites/APIs without caching support wouldn't have ETags, so would be nice to offer _something_. Did you see my fallback idea in https://github.com/concourse/archive-resource/issues/4? Ditto about the "don't worry about it for this pull request".

**Comment by DavidEBest** [comment]
Added a comment about the two types of versioning currently available. I want to try and get this merged upstream - if the approach is validated, then I'll add checksum-based versioning next.

**Comment by DavidEBest** [comment]
Upstream PR here: https://github.com/aequitas/concourse-http-resource/pull/1
